### DISCUSSION TITLE
a function to fetch a python file and automatically fetch (relative import) used modules, (optional)  run them once

### DISCUSSION THREAD

**Discussion by grimmerk** [pyodide/pyodide#1876] [Ideas]
From my perspective, there are two applications of Pyodide, one is REPL, the other one is to load Python script(s) to run and it could be a small piece of a code or a larger codebase. I have migrated my [Chrome extension](https://chrome.google.com/webstore/detail/dicom-image-viewer/ehppmcooahfnlfhhcflpkcjmonkoindc) to Pyodide base from pure JS, and been published. Thank you for this great tool. I may want to refactor my codes to multiple files and add more functionalities. It would be great if there is such a feature/API introduced in the title. With it, I do not need to fetch one by one Python files and run them. 

Supporting the overall Python native import system may be a difficult/long-term task, and starting from "relative import" or even limit to the same folder might be easier. I'm not sure whether this approach is good or not.

**Comment by ryanking13** [comment]
I agree. Though we can fetch external files then execute it, people will simply want to `import module`.

For that, I think we may extend import system such as:

- Fetch external python files,
- Save it to virtual file system,
- Then import it

The main problem of this approach will be how to properly fetch external python files that are import candidates
and verify that they are properly downloaded.

**Reply by rth** [reply]
On a related topic there was some initial work to implement to `<script type="text/python">` (which can be extended to `<script type="text/python" src="...python.py">` in https://github.com/pyodide/pyodide/pull/692 which might make some of this easier if I understand your request correctly.  

>> Supporting the overall Python native import system may be a difficult/long-term task,
>
> For that, I think we may extend import system such as:

Yes, I'm a bit weary of trying to hack the python import mechanisms to auto-download packages. In the code directly provided by the user as we are doing for Pyodide packages, that might work as opt-in, in some limited cases. 
However otherwise Python has plenty of `try: import module except: do something else` for instance for left overs of Py2/Py3 compatibility and if there is an HTTP request happening for each of those that would be very problematic. That would render import times completely non deterministic.  Also it's pretty common to make typo in imports, which in this case would try to install arbitrary packages.

See related discussions in https://github.com/pyodide/pyodide/issues/1470

**Reply by rth** [reply]
I also think that we don't necessarily need to re-invent the wheel. In a lot of cases, it's interesting to see if such a feature exists in JavaScript (in this case auto-download of relatively imported modules), if yes -- use it as inspiration, and if not -- investigate why not.

In this case, if I understand your use case, the idea is that we have some number of files, say on Github and we would like to load them directly from a single URL with some relative imports, without specifying all the paths ?  Can we do that assuming those were JS files though (true question -- I don't know the JS ecosystem well)? Without using a bundler (or in our case creating a Python very wheel), or modifying the script tags in DOM which would be our equivalent of calling micropip maybe.

**Reply by hoodmane** [reply]
> it's interesting to see if such a feature exists in JavaScript

Most related thing is [import maps](https://github.com/WICG/import-maps). (Import maps has been released in Chrome. Firefox is planning to implement and has a prototype.)

**Reply by rth** [reply]
Very interesting thanks! That's exactly the kind of document that I meant.

In the end the solution that they came up with looks more like a kind of a requirement file. Or in their case a .importmap file that can be included inline in a `script` element. It's possible to generate [the contents dynamically](https://github.com/WICG/import-maps#dynamic-import-map-example) however this must happen before any of the imports actually happen.

Also worth reading the [alternative solutions](https://github.com/WICG/import-maps#alternatives-considered) they have evaluated and in particular,

> ### The Node.js module resolution algorithm
>
> Unlike in Node.js, in the browser we don't have the luxury of a reasonably-fast file system that we can crawl looking for modules. Thus, we cannot implement the [Node module resolution algorithm](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) directly; it would require performing multiple server round-trips for every import statement, wasting bandwidth and time as we continue to get 404s. We need to ensure that every import statement causes only one HTTP request; this necessitates some measure of precomputation.
>
> ### A programmable resolution hook
>
> Some have suggested customizing the browser's module resolution algorithm using a JavaScript hook to interpret each module specifier.
> 
> Unfortunately, this is fatal to performance; jumping into and back out of JavaScript for every edge of a module graph drastically slows down application startup.

Maybe the conclusions would be different for Python but my guess is that they wouldn't be.

On the other hand I think implementing something like their importmap for Pyodide would be really nice.
```html
<script type="pyimportmap">
{
  "imports": {
    "foo": "https://some.example/foo.mjs",
    "bar": "https://other.example/bar.mjs"
  }
}
</script>
```
which is more general that the `type="text/python"` that we considered before, as those modules can then be copied inside the virtual file-system and imported as needed. Though we would probably still need some way to put inline Python code in HTML, as other thing than JS string. It also remains to be determined how to make this work with existing say PyPi or Pyodide packages and dependency resolution.

I read the spec quickly, and I'm not sure I understand their URL remapping approach though.

WDYT? Would such import maps address your use case for instance @grimmer0125 ?

**Reply by hoodmane** [reply]
> Maybe the conclusions would be different for Python but my guess is that they wouldn't be.

In my opinion, the first conclusion still applies. I think the conclusion for "A Programmable Resolution Hook" does NOT apply, because Python already comes with programmable module resolution hooks so whatever the costs of such a hook are we are already paying. 

That said, I am personally in favor of something like import maps because I think it satisfies the need well and is quite simple. 

In any case, the main problem is how to block for the `fetch`. I suppose we can use synchronous `XMLHttpRequest` but this is a bit of an anti-pattern. Especially if we want to fetch zip archives, it would be a drag to have to base 64 encode the data to deal with the fact that synchronous XHR doesn't support binary data.

**Reply by hoodmane** [reply]
Also, because Python already has dynamic import hooks, we don't necessarily need the restriction that the import map is established before loading the runtime.

**Reply by ryanking13** [reply]
> Yes, I'm a bit weary of trying to hack the python import mechanisms to auto-download packages. In the code directly provided by the user as we are doing for Pyodide packages, that might work as opt-in, in some limited cases.

I am trying to implement auto-download mechanism in this way, you can see the WIP code [here](https://github.com/ryanking13/pyodide/blob/import-support/src/py/pyodide/importer.py).

> However otherwise Python has plenty of try: import module except: do something else for instance for left overs of Py2/Py3 compatibility and if there is an HTTP request happening for each of those that would be very problematic.

It's true that in this way, all "failed imports" will generate at least two HTTP requests (\<module>.py, \<module>/\_\_init__.py).

To prevent this, I think we can try something like "register before import" approach which [httpimport](https://github.com/operatorequals/httpimport) uses.

If this doesn't seem totally bad, I'll make a issue for a proposal.

**Reply by hoodmane** [reply]
> If this doesn't seem totally bad

I don't like it very much. My opinion is that this feature should probably wait on synchronous I/O so we can handle binary files. I would also prefer a design that includes some sort of package manifest file so that always-necessary parts of the package can be downloaded together. As is a package with a large number of files will take forever to import.

**Reply by ryanking13** [reply]
> My opinion is that this feature should probably wait on synchronous I/O so we can handle binary files.

True. I couldn't find ways to support downloading binary files synchronously such as .pyc or .zip for now.

> I would also prefer a design that includes some sort of package manifest file so that always-necessary parts of the package can be downloaded together.

Something like the `packages.json` file we use to import Pyodide packages like numpy?
I think python files can be dynamically added on runtime for many use cases like browser-based online editor (#1784), which is (probably) not possible with the manifest file.

**Reply by hoodmane** [reply]
For the manifest file, I was thinking along the lines that we could record what set of files were needed in response to a given import the first time and then put that information somewhere so that on later imports the files could be prefetched. I am not sure if this is actually a good idea though.

**Reply by ryanking13** [reply]
How do you think of having a global variable that can be updated dynamically instead of a manifest file?

For example, `pyodide.importable_modules = ["package1", "package2"]`, which keeps the top package names or all sub-package names that can be imported externally. (Sorry about the poor naming 😂.)

Actually, this approach cannot prevent importing a large package from taking a long time to be downloaded,
but prefetching packages will also consume time at some stage (maybe at the initialization?).

BTW, If someone is trying to use a Python package with a bunch of files, it will be better to recommend him/her to make a standalone whl package for it, instead of fetching each file individually.

**Reply by grimmerk** [reply]
hi, I'm sorry that I did not follow up on the bove discussion and just take a look quickly.  The usage scenario of my original idea is 
1) just an API to call like `await pyodide.fetchAndRunPythonAsync(http://localhost/1.py)` and it will automatically download the used user-defined and imported files, e.g. `http://localhost/2.py` used in `1.py`.  
2) This is mainly for local usage, which means front-end code and the Python files are on the **same** machine/PC/server, so using **relative Python import** to achieve automatical importing might be convenient thing.  For those cases which fetch remote Python files on a non-local http_url (e.g. GitHub), I think they are kind of deployed files so that using **wheel** is more acceptable than previoius case. 
3)  Looks like there are 3 possible designs 
    1. current: multiple `const pythonCode = await (await fetch(url)).text();` + `const num = awaitpyodide.runPythonAsync(python_code_1)`
    2. my suggested ideal API: users just need to use 1 API and does not do any extra config
    3. extra a config mentioned in the above discussion (I hope I understand the above discussion correctly), like below and trigger some API to load them once. 
    ```
    "imports": {
      "foo": "https://some.example/foo.py",
      "bar": "https://other.example/bar.py"
    }
    ``` 
    I think 2 is the most convenient one (regardless of potential issues) but 3. might be already better than 1 since it is systematic and less duplicated API call than 1.

**Reply by ryanking13** [reply]
> This is mainly for local usage, which means front-end code and the Python files are on the same machine/PC/server, so using relative Python import to achieve automatical importing might be convenient thing. For those cases which fetch remote Python files on a non-local http_url (e.g. GitHub), I think they are kind of deployed files so that using wheel is more acceptable than previoius case.

I'm thinking of the same usage. Sorry if my sentences were misleading due to my bad english skills 👀.

Basically what I expect is similar to the approach of [Brython import system](https://brython.info/static_doc/en/import.html).

**Reply by hoodmane** [reply]
So one thing that could be said about `fetchAndRunPythonAsync` is that it can be easily defined by an end user that wants it:
```js
async fetchAndRunPythonAsync(url) {
   let resp = await fetch(url);
   if(!resp.ok){
      throw new Error("Better error message here");
   }
   let source = await resp.text();
   return await runPythonAsync(source);
}
```
Of course just because it is easily defined doesn't mean we shouldn't include it but it's worth considering whether such an API would be helpful. Another option would be to advertise this definition in the FAQ first.

**Reply by hoodmane** [reply]
Another option would be an API to download a list of files and place them in a particular directory. Maybe something like
```js
pyodide.installFiles(base_url, install_directory, file_list)
```
maybe used like:
```js
await pyodide.installFiles("https://mywebsite.com/resources/my_py_lib/src", "/my_py_lib",
   ["__init__.py", "some_module.py", "subfolder/item_a.py", "subfolder/item_b.py"]
);
pyodide.runPython("from my_py_lib import do_the_thing; do_the_thing()");
```
We could suggest for people to create a manifest using some command line tool, in which case they could instead do:
```js
let files = await (await fetch("https://mywebsite.com/resources/my_py_lib/manifest.json)).json();
await pyodide.installFiles("https://mywebsite.com/resources/my_py_lib/src", "/my_py_lib", files);
```
Or maybe if installing a zip file would be more useful:
```js
await pyodide.installZipArchive("https://mywebsite.com/resources/my_py_lib/my_py_lib.zip", "/my_py_lib");
```
I think this is actually closer to what people need.

**Reply by hoodmane** [reply]
> This is mainly for local usage

Ah right, for getting local development code working quickly something like what @ryanking13 was making is actually a pretty decent idea. For a deployed application it's an abomination.

Here's a thought: what if we added `installZipArchive` or `installFiles` or something like that for peoples' deployed apps, and @ryanking13 could add his import hook either as a PyPi package or as Pyodide package. Then people could import that package on dev setups. We can put a warning like "really please only use this on localhost".

**Reply by ryanking13** [reply]
> add his import hook either as a PyPi package or as Pyodide package

Alright, I'll try making it for a Pyodide package, since it will require js fetch (and readFileSync maybe for node).

So if I understanded correctly, we are going to provide three ways to use external Python files:

1. `micropip.install` - for standalone, pure python, and non-Pyodide dependent packages.
2. `Pyodide package with hacks Python import hook` - which supports relative import mostly for local dev setups.
3. `pyodide.installFiles` or `pyodide.installZipArchive` - which supports relative import mostly for production setups.

After the implementation is finished, we will need to provide documentation for this, something like _"Modularizing python code"_ section.

**Reply by hoodmane** [reply]
> I'll try making it for a Pyodide package, since it will require js fetch (and readFileSync maybe for node).

It is perfectly possible to stick a Pyodide-only package on PyPi. Dunno if it is discouraged. I did that with [pyodide-interrupts](https://pypi.org/project/pyodide-interrupts/) which is possible to import and use outside of Pyodide but also kind of useless. Though it's also useless to use in the currently version of Pyodide because we added the keyboard interrupt system.

But probably the best approach is indeed to include it as part of Pyodide.

> we will need to provide documentation for this

The documentation is arguably the most important step.

**Reply by rth** [reply]
Maybe let's open separate issues to discuss points 2 and 3 and how the API should look before implementing it? The collapsible threads in "discussions" don't seem designed for long discussions.

> Pyodide package with hacks Python import hook

Wouldn't it be easier to make a PyPi package for it, as a start?

> pyodide.installFiles

Isn't this somewhat what the importmap system is about? We can certainly come up with our or own API, but it would need more design work to make sure it's right (and to avoid more confusion with existing installation methods). What I like about about `installFiles` is that it could be used on other files than Python ones though.

**Reply by rth** [reply]
> It is perfectly possible to stick a Pyodide-only package on PyPi. Dunno if it is discouraged.

I think we should encourage putting packages to PyPi rather than vendoring them in Pyodide, when possible. Particularly for experimental stuff, it allows to move and iterate much faster (no need to wait for a Pyodide release for a bugfix). There are plenty of packages that don't make sense by themselves, all the `pytest-*` plugins, `django-*` plugins etc. From a Python package point of view all Pyodide does is to expose a few extra imports (e.g. `js`) which is fairly common. So as long as the name is pre-fixed with `pyodide-` I think it should be fine. We should also have a list of such packages somewhere.

**Reply by rth** [reply]
>  name is pre-fixed with pyodide-

BTW if you are interested @ryanking13 we could also host that future package (once it's ready) at the Pyodide Github org, so it's a bit easier to discover.

**Reply by ryanking13** [reply]
> Maybe let's open separate issues to discuss points 2 and 3 and how the API should look before implementing it?

I made a proposal to discuss option 2. #1917

> we could also host that future package (once it's ready) at the Pyodide Github org

Sure, I would very much appreciate that!

**Comment by grimmerk** [comment]
> Also worth reading the [alternative solutions](https://github.com/WICG/import-maps#alternatives-considered) they have evaluated and in particular
...wasting bandwidth and time...

I think it might not be a big deal for small or medium size of a group of Python files. But it indeed needs to do some time measurement for any new feature for module resolution/hook/hack.   

@ryanking13  [Brython import system](https://brython.info/static_doc/en/import.html) looks good. 

--

Let me explain more about the initial proposal. 
- I saw `pyodide.loadPackagesFromImports` and guess it already implement some kind of detect/scan the used `import statement` list in one python script string. So this is one of the reasons I propose to do a similar but more complicated thing. 
- > _await pyodide.fetchAndRunPythonAsync(http://localhost/1.py) and it will automatically download the used user-defined and imported files, e.g. http://localhost/2.py used in 1.py._

    Originally what I imagine is a user defines `from .module2 import sample_func` in `module1.py` file, Pyodide can detect this and fetch `module2` python file. There are two benefit
    1. It is straightforward and uses the same import syntax as the original Python module resolution. 
    2. If the Python scripts do not use any `Pyodide` specific Python API inside, people can use local Python to do any tests for multiple Python files as they use the same import syntax. But I realize this case is not a common case when using Pyodide. So this benefit is not very much.
- Regarding `wheel` or different user scenarios about `dev/deploy/production` environments. I may not clarify my ideas in my previous comment very well. There are 3 cases
    1. For local development,  the proposed feature (auto-detect/fetch) in this discussion is helpful, or alternatively, some auto-live-bundle wheel mechanisms may have the same benefit as people need to see the result from the code change sooner. 
    2. To deploy a web project composed of JavaScript files and Pyodide scripts to some cloud, they often belong to the same GitHub repo (mono repo), personally, I think packaging Python scripts as a wheel may be acceptable. However, packaging as a wheel is still not straightforward for a self-own project, instead of shipping a library. Also, it needs some effort to set up it if using traditional setuptool. [Poetry build](https://python-poetry.org/) almost does not pay any effort but it has some limitations on your project folder structure, https://github.com/python-poetry/poetry/issues/2252, which requires people to put codes in a first level folder and name the folder as the project name in `pyproject.toml`.
    3. To deploy a web project composed of JavaScript files which use some Python scripts belonging to another project for Pyodide usage, in this case, these Python files are not in the same cloud server, and packaging them as wheels seem very reasonable.

- > We can put a warning like "really please only use this on localhost".

    @hoodmane, 
It seems that you have some concerns to apply this (implicit auto module resolution) other than local dev. It may be risky/unstable and encounter some `http/resolution` error so that it is hard to give the same guarantee as `pyodide.installFiles` design which uses an explicit file list instead of module resolution. Is this your concern? This is just a confirmation with you to avoid my incorrect understanding since I'm not involved in the above discussion. If so, I agree that this kind of feature may not become stable in a short time, using Python `wheels` is proven.

**Comment by grimmerk** [comment]
hi, I notice Python package management `Poetry`has solved the mentioned issue above, https://github.com/python-poetry/poetry/issues/2252. Although it is not directly related to this topic, it may be still valuable to share.

in the generated `pyproject.toml`, you can customize the packages folder name and location, then they can differ to the default project/package , e.g. put your code inside a `abcd` folder, abc is the new package name.  

```toml
[tool.poetry]
name = "pyodide-helper"

packages = [
    { include = "abcd", from = "." }
]
```

Then `from abcd import test`.

ref: https://github.com/python-poetry/poetry/pull/3257/files

This simplifies the building wheel setup.
